File: mvp_app/MyProject/libtsparser/src/main/java/com/tosmart/xiexuming/libtsparser/table/ServiceDescriptionTableManager.java
package com.tosmart.xiexuming.libtsparser.table;


import com.tosmart.xiexuming.libtsparser.descriptor.Descriptor;
import com.tosmart.xiexuming.libtsparser.descriptor.DescriptorManager;
import com.tosmart.xiexuming.libtsparser.section.ServiceDescriptionSection;

import java.util.ArrayList;
import java.util.HashMap;

/**
 * @author xiexuming
 */
public class ServiceDescriptionTableManager {
    private static final int SERVICE_HEAD_LENGTH = 5;

    private final HashMap<Integer, ServiceDescriptionSection> mSections = new HashMap<>();
    private int mVersionNumber = -1;

    public boolean composeSection(ServiceDescriptionSection serviceDescriptionSection) {
        if (serviceDescriptionSection == null) {
            return false;
        }
        int lastSectionNumber = serviceDescriptionSection.getLastSectionNumber();
        if (mVersionNumber != -1) {
            if (mVersionNumber != serviceDescriptionSection.getVersionNumber()) {
                mVersionNumber = -1;
                mSections.clear();
                return false;
            }
            if (mSections.containsKey(serviceDescriptionSection.getVersionNumber())) {
                return false;
            }
        }
        mSections.put(serviceDescriptionSection.getSectionNumber(), serviceDescriptionSection);
        mVersionNumber = serviceDescriptionSection.getVersionNumber();
        return mSections.size() == (lastSectionNumber + 1);
    }

    public ServiceDescriptionTable getServiceDescriptionTable() {
        if (mSections.size() <= 0) {
            return null;
        }
        ServiceDescriptionTable serviceDescriptionTable = new ServiceDescriptionTable();
        for (int i = 0; i < mSections.size(); i++) {
            ServiceDescriptionSection serviceDescriptionSection = mSections.get(i);
            byte[] serviceData = serviceDescriptionSection != null ? serviceDescriptionSection.getData() : null;
            if (serviceData != null) {
                ArrayList<Service> services = composeServices(serviceData);
                serviceDescriptionTable.setService(services);
            }
        }
        serviceDescriptionTable.setVersionNumber(mVersionNumber);
        return serviceDescriptionTable;
    }

    private ArrayList<Service> composeServices(byte[] serviceData) {
        if (serviceData.length <= 0) {
            return null;
        }
        ArrayList<Service> serviceList = new ArrayList<>();
        for (int i = 0; i < serviceData.length; ) {
            Service service = new Service();
            int serviceId = (serviceData[i] & 0xFF) << 8 | (serviceData[i + 1] & 0xFF);
            service.setServiceId(serviceId);
            int scheduleFlag = serviceData[i + 2] >> 1 & 0x01;
            service.setEITScheduleFlag(scheduleFlag);
            int presentFollowingFlag = serviceData[i + 2] & 0x01;
            service.setEITPresentFollowingFlag(presentFollowingFlag);
            int runningStatus = (serviceData[i + 3] & 0xE0) >> 5;
            service.setRunningStatus(runningStatus);
            int freeMode = (serviceData[i + 3] & 0x10) >> 4;
            service.setFreeCAMode(freeMode);
            int descriptorsLoopLength = (serviceData[i + 3] & 0x0F) << 8 | serviceData[i + 4];
            service.setDescriptorsLoopLength(descriptorsLoopLength);

            if (descriptorsLoopLength > 0) {
                byte[] descriptorData = new byte[descriptorsLoopLength];
                int serviceDescriptorStart = i + SERVICE_HEAD_LENGTH;
                System.arraycopy(serviceData, serviceDescriptorStart, descriptorData, 0, descriptorsLoopLength);
                ArrayList<Descriptor> descriptors = DescriptorManager.composeDescriptor(descriptorData);
                service.setDescriptor(descriptors);
            }

            serviceList.add(service);
            i = i + SERVICE_HEAD_LENGTH + descriptorsLoopLength;
        }
        return serviceList;
    }
}

File: newpaserapp/libtosmartparse/src/main/java/com/example/libtosmartparse/table/ServiceDescriptionTableManager.kt
package com.example.libtosmartparse.table

import com.example.libtosmartparse.descriptor.Descriptor
import com.example.libtosmartparse.descriptor.DescriptorManager
import com.example.libtosmartparse.packet.or
import com.example.libtosmartparse.packet.shl
import com.example.libtosmartparse.packet.shr
import com.example.libtosmartparse.section.ServiceDescriptionSection
import kotlin.experimental.and

/**
 * @author xxm
 * 2022/7/17
 **/
class ServiceDescriptionTableManager {
    private val mSections: HashMap<Int, ServiceDescriptionSection> =
        HashMap<Int, ServiceDescriptionSection>()
    private var mVersionNumber = -1
    fun composeSection(serviceDescriptionSection: ServiceDescriptionSection?): Boolean {
        if (serviceDescriptionSection == null) {
            return false
        }
        val lastSectionNumber: Int = serviceDescriptionSection.getLastSectionNumber()
        if (mVersionNumber != -1) {
            if (mVersionNumber != serviceDescriptionSection.getVersionNumber()) {
                mVersionNumber = -1
                mSections.clear()
                return false
            }
            if (mSections.containsKey(serviceDescriptionSection.getVersionNumber())) {
                return false
            }
        }
        mSections[serviceDescriptionSection.getSectionNumber()] = serviceDescriptionSection
        mVersionNumber = serviceDescriptionSection.getVersionNumber()
        return mSections.size == lastSectionNumber + 1
    }

    fun getServiceDescriptionTable(): ServiceDescriptionTable? {
        if (mSections.size <= 0) {
            return null
        }
        val serviceDescriptionTable = ServiceDescriptionTable()
        for (i in 0 until mSections.size) {
            val serviceDescriptionSection: ServiceDescriptionSection? = mSections[i]
            val serviceData: ByteArray? =
                if (serviceDescriptionSection != null) serviceDescriptionSection.getData() else null
            if (serviceData != null) {
                val services = composeServices(serviceData)
                serviceDescriptionTable.setService(services)
            }
        }
        serviceDescriptionTable.setVersionNumber(mVersionNumber)
        return serviceDescriptionTable
    }

    private fun composeServices(serviceData: ByteArray): ArrayList<Service>? {
        if (serviceData.size <= 0) {
            return null
        }
        val serviceList = ArrayList<Service>()
        var i = 0
        while (i < serviceData.size) {
            val service = Service()
            //programData[j] shl 8 or programData[j + 1].toInt() and 0xFF
            val serviceId: Int = serviceData[i]  shl 8 or serviceData[i + 1].toInt() and 0xFF
            service.setServiceId(serviceId)
            val scheduleFlag: Int = serviceData[i + 2] shr 1 and 0x01
            service.setEITScheduleFlag(scheduleFlag)
            val presentFollowingFlag: Byte = serviceData[i + 2] and 0x01
            service.setEITPresentFollowingFlag(presentFollowingFlag.toInt())
            val runningStatus: Int = serviceData[i + 3] and 0xE0.toByte() shr 5
            service.setRunningStatus(runningStatus)
            val freeMode: Int = serviceData[i + 3] and 0x10 shr 4
            service.setFreeCAMode(freeMode)
            val descriptorsLoopLength: Int = serviceData[i + 3] and 0x0F shl 8 or serviceData[i + 4]
            service.setDescriptorsLoopLength(descriptorsLoopLength)
            if (descriptorsLoopLength > 0) {
                val descriptorData = ByteArray(descriptorsLoopLength)
                val serviceDescriptorStart = i + SERVICE_HEAD_LENGTH
                System.arraycopy(
                    serviceData,
                    serviceDescriptorStart,
                    descriptorData,
                    0,
                    descriptorsLoopLength
                )
                val descriptors: ArrayList<Descriptor> =
                    DescriptorManager.composeDescriptor(descriptorData)!!
                service.setDescriptor(descriptors)
            }
            serviceList.add(service)
            i = i + SERVICE_HEAD_LENGTH + descriptorsLoopLength
        }
        return serviceList
    }

    companion object {
        private const val SERVICE_HEAD_LENGTH = 5
    }
}
